src/v1.js

```diff
@@ -567,55 +575,66 @@ api.declare({
     return;
   }
 
+  input.scopes.sort(scopeCompare);
+
+  let when = new Date().toJSON();
+  role = {
+    roleId,
+    description: input.description,
+    scopes: input.scopes,
+    lastModified: when,
+    created: when,
+  };
+
+  // update Roles
+  let reportError = (code, message, details) => {
+    res.reportError(code, message, details);
+    let err = new Error();
+    err.code = 'ErrorReported';
+    return err;
+  };
   try {
-    this.resolver.checkUpdatedRole(roleId, input.scopes);
-  } catch (e) {
-    return res.reportError('InputError', e.message, {});
-  }
-
-  let role = await this.Role.create({
-    roleId:       roleId,
-    description:  input.description,
-    scopes:       input.scopes,
-    details: {
-      created:      new Date().toJSON(),
-      lastModified: new Date().toJSON(),
-    },
-  }).catch(async (err) => {
-    // Only handle
-    if (err.code !== 'EntityAlreadyExists') {
-      throw err;
-    }
+    await this.Roles.modify(roles => {
+      let existing = _.find(roles, {roleId});
+      if (existing) {
+        // role exists and doesn't match this one -> RequestConflict
+        if (existing.description !== input.description || !_.isEqual(existing.scopes, input.scopes)) {
+          throw reportError('RequestConflict',
+            'Role with same roleId already exists',
+            {});
+        } else {
+          role = existing;
+          return;
+        }
+      }
 
-    // Load role
-    let role = await this.Role.load({roleId});
+      // check that this new role does not introduce a cycle
+      let checkRoles = _.clone(roles);
+      checkRoles.push(role);
+      try {
+        ScopeResolver.cycleCheck(checkRoles);
+      } catch (e) {
+        throw reportError('InputError', `Invalid roles: ${e.message}`, {});
+      }
 
-    // If stored role different or older than 15 min we return 409
-    let created = new Date(role.details.created).getTime();
-    if (role.description !== input.description ||
-        !_.isEqual(role.scopes, input.scopes) ||
-        role > Date.now() - 15 * 60 * 1000) {
```

Comment:
note that the old code here compared the role object to a Date object..

Reply:
nice catch